cluster/mesos/cluster.go

```diff
@@ -0,0 +1,453 @@
+		for _, offer := range s.offers {
+			list = append(list, offer)
+		}
+	}
+	return list
+}
+
+// Info gives minimal information about containers and resources on the mesos cluster
+func (c *Cluster) Info() [][2]string {
+	offers := c.listOffers()
+	info := [][2]string{
+		{"\bStrategy", c.scheduler.Strategy()},
+		{"\bFilters", c.scheduler.Filters()},
+		{"\bOffers", fmt.Sprintf("%d", len(offers))},
+	}
+
+	sort.Sort(offerSorter(offers))
+
+	for _, offer := range offers {
+		info = append(info, [2]string{" Offer", offer.Id.GetValue()})
+		for _, resource := range offer.Resources {
+			info = append(info, [2]string{"  └ " + *resource.Name, fmt.Sprintf("%v", resource)})
+		}
+	}
+
+	return info
+}
+
+func (c *Cluster) addOffer(offer *mesosproto.Offer) {
+	s, ok := c.slaves[offer.SlaveId.GetValue()]
+	if !ok {
+		return
+	}
+	s.addOffer(offer)
+	go func(offer *mesosproto.Offer) {
+		<-time.After(c.offerTimeout)
+		if c.removeOffer(offer) {
+			if _, err := c.driver.DeclineOffer(offer.Id, &mesosproto.Filters{}); err != nil {
+				log.WithFields(log.Fields{"name": "mesos"}).Errorf("Error while declining offer %q: %v", offer.Id.GetValue(), err)
+			} else {
+				log.WithFields(log.Fields{"name": "mesos"}).Debugf("Offer %q declined successfully", offer.Id.GetValue())
+			}
+		}
+	}(offer)
+}
+
+func (c *Cluster) removeOffer(offer *mesosproto.Offer) bool {
+	log.WithFields(log.Fields{"name": "mesos", "offerID": offer.Id.String()}).Debug("Removing offer")
+	s, ok := c.slaves[offer.SlaveId.GetValue()]
+	if !ok {
+		return false
+	}
+	found := s.removeOffer(offer.Id.GetValue())
+	if s.empty() {
+		// Disconnect from engine
+		delete(c.slaves, offer.SlaveId.GetValue())
+	}
+	return found
+}
+
+func (c *Cluster) scheduleTask(t *task) bool {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
+
+	n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), t.config)
+	if err != nil {
+		return false
+	}
+	s, ok := c.slaves[n.ID]
+	if !ok {
+		t.error <- fmt.Errorf("Unable to create on slave %q", n.ID)
+		return true
+	}
+
+	// build the offer from it's internal config and set the slaveID
+	t.build(n.ID)
+
+	c.Lock()
+	// TODO: Only use the offer we need
+	offerIds := []*mesosproto.OfferID{}
+	for _, offer := range c.slaves[n.ID].offers {
+		offerIds = append(offerIds, offer.Id)
+	}
+
+	if _, err := c.driver.LaunchTasks(offerIds, []*mesosproto.TaskInfo{&t.TaskInfo}, &mesosproto.Filters{}); err != nil {
+		// TODO: Do not erase all the offers, only the one used
+		for _, offer := range s.offers {
+			c.removeOffer(offer)
+		}
+		s.Unlock()
+		t.error <- err
+		return true
+	}
+
+	s.addTask(t)
+
+	// TODO: Do not erase all the offers, only the one used
+	for _, offer := range s.offers {
+		c.removeOffer(offer)
+	}
+	c.Unlock()
+	// block until we get the container
+	finished, err := t.monitor()
+	taskID := t.TaskInfo.TaskId.GetValue()
+	if err != nil {
+		//remove task
+		s.removeTask(taskID)
+		t.error <- err
+		return true
+	}
+	if !finished {
+		go func() {
```

Comment:
What's the purpose of this?

It waits until its finished then does nothing, right? I assume it's preliminary code and sometime it's going to do something else. Would you mind adding a comment?